src/key_scan.rs

```diff
@@ -89,31 +144,40 @@ pub enum ScanStatus<L: Ledger> {
 /// recipient. It is able to create Merkle paths for the new records either by using the Merkle
 /// paths included in `Memos` events, or by using the frontier at the time it appends the
 /// commitment for an attached record opening to the tree.
-#[ser_test(arbitrary, ark(false), types(reef::cap::Ledger))]
+#[ser_test(
+    arbitrary,
+    ark(false),
+    types(reef::cap::Ledger, ViewerKeyPair),
+    types(reef::cap::Ledger, FreezerKeyPair),
+    types(reef::cap::Ledger, UserKeyPair)
+)]
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(bound = "L: Ledger")]
-pub struct BackgroundKeyScan<L: Ledger> {
-    key: UserKeyPair,
+pub struct BackgroundKeyScan<L: Ledger, Key: KeyPair + DeserializeOwned + Serialize> {
+    key: Key,
     // The index of the next event in the event stream.
     next_event: EventIndex,
     // The first event of interest.
     from_event: EventIndex,
     // The first event after the range of interest.
     to_event: EventIndex,
-    // Record openings we have discovered which belong to these key. These records are kept in a
-    // separate pool until the scan is complete so that if the scan encounters an event which spends
-    // some of these records, we can remove the spent records without ever reflecting them in the
-    // keystore's balance.
+    // Record openings we have discovered which are viewable, freezable, or belong to these key.
+    // These records are kept in a separate pool until the scan is complete so that if the scan
+    // encounters an event which spends some of these records, we can remove the spent records
+    // without ever reflecting them in the keystore's balance.
     records: HashMap<Nullifier, (RecordOpening, u64)>,
```

Comment:
If I read the code correctly we actually only add to this if the key in scan is `Sending`

Reply:
Yeah, only the scan for sending is added to `records` since we don't have the nullifier for viewing or freezing. I think I should change `records`. It should be `HashMap` only for sending, and `HashSet<(RecordOpening, u64)>` for viewing and freezing.